### PR TITLE
Brownian unification

### PR DESCRIPTION
**Pull request by patrick-kidger**
As per #60. (Also obsoletes #54, a great many things in #15, and a bit of #22)

Merged BrownianPath/Tree/Interval. BrownianPath and Tree are wrappers for BrownianInterval with particular arguments. In particular this means all of them are now on the same interface, providing the same functionality, and of course this is much easier to support. Plus, no C++ or blist cluttering things up.

This also includes several improvements to BrownianInterval:
- Use of trampolining and tail recursion
- Fewer SeedSequences are now spawned
- Those SeedSequences that are spawned now use a more efficient `spawn_key`, which should resolve the previous issue that tuple additions were taking up an appreciable fraction of the run time. The spawn keys are now a 2-tuple of `(depth in tree, index at depth level)`, rather than being long tuples that get longer each time we split.

There are some small downsides:
- We're no longer as efficient for point-based queries. The efficiency drop should be _relatively_ small: the increase in cost is logarithmic in `(length of the interval / average step size)`, and moreover this doesn't affect us in torchsde anyway as we only ever made interval-based queries.
- We're not longer running at C++ speeds. Practically speaking the difference doesn't seem to be huge: on the `sdeint_adjoint` benchmark,
    Old+Path+CPP=3.2s vs New+Path+Py=4s.
    Old+Tree+CPP=5s vs New+Tree+Py=8s.
It's also interesting to see some of the Python speedups:
    Old+Tree+Py=25s vs New+Tree+Py=8s.
    Old+Interval+Py=13s vs New+Tree+Py=7.8s
(Incidentally this is Tree w/ tol=1e-5 rather than the default 1e-6.)
If we do want to return to C++ then I'd suggest using the current BrownianInterval code as a model.

Overall I definitely think the upsides outweight the downsides.

### CODE REVIEW

**Comment by lxuechen** [comment]
Thanks for the update. 

This PR is longer than I expected, and I'll try my best to go through it in detail soon, but it's possible that it may come after the ICLR deadline. 

A few non-ideal changes I notice already are
1. the breaking of backwards compatibility, and
2. the removal of the Brownian base class. 

1 is typically something that drives people mad, so it's essential that we support the old API. 2 is also not ideal in the long run, IMHO. A base class should help tell us and people what we expect the Brownian stuff to do, and I kinda view it as a header file that defines the whole API. Keeping it there would help other smart people potentially extend this if they wished, just like when you came up with `BrownianInterval`.

**Comment by patrick-kidger** [comment]
1. Ah, I'd assumed from the removal of logqp that we were allowing backward-incompatible changes. I see now that you're planning on including this anyway, so sure, I can make things backward-compatible.
2. Fair point.

I'll make the changes.

**Comment by patrick-kidger** [comment]
Okay, I've added BaseBrownian.

Regarding BrownianPath, BrownianTree: it's not feasible to have these simultaneously (a) be convenience wrappers for BrownianInterval with different default arguments, and (b) support the old API. It sounds like you prefer (b) to (a).

As such I see two possible options:
- Implement BPath and BTree by wrapping BInterval, which produces very minor backward-incompatible changes (for BPath, can no longer query outside [t0, t1]; `.to` would no longer function).
- Copy-paste all the old BPath/BTree code in from master, additionally switching out blist for a dictionary to keep functionality if not speed.

In either case adding an additional warning that BPath/BTree is deprecated and should be switched out for BInterval with appropriate arguments.

Does that fit with what you're thinking?

**Comment by lxuechen** [comment]
Thanks for adding the base class back. 

I think my main point about backwards comp. is about the constructor method `__init__`. It would be nice if people could create BPath instances as before, with the same set of arguments. 

I agree to deprecate `.to`, and let the interpreter raise a NotImplError stating this particular method is deprecated when called. 

I agree with the wrapper, but don't agree with raising additional warnings that BPath and BTree is deprecated, as they shouldn't be. At the very least, BPath should be a core part of the brownian component, after I patch changes to the C++ code. This is by far a very easy way to get the speed up for small problems, and I don't see how 3.2s vs 4s is not a big regression in terms of speed. 

Happy to discuss more on points where we disagree.

**Comment by patrick-kidger** [comment]
Okay, so, plan is to have BPath, BTree be wrappers for BInterval, using the old API. This will result in minor backward incompatibilities (no `.to`) but should be fine. The main advantage being that there is now only have one Brownian* object to maintain, rather than five.

- The reason for the deprecation warnings is because _not_ deprecating them implies that we're happy with the old APIs, which are (a) inconsistent with BInterval, (b) a sharply limited subset of what BInterval is capable of, and (c) involve a couple of arguments that now do nothing, e.g. `window_size` on BPath. To my mind it's better to point people towards the better thing to be using, i.e. to favour `BInterval(..., cache_size=None)` over BPath.
- The speed regression is really just because of the C++ -> Python switch, not because of the choice to implement BPath via BInterval. As such I anticipate that it's only temporary. My suggestion for C++ going forwards is to implement _just_ BrownianInterval.
*Footnote - to me, the speed improvement from 4s to 3.2s is not worth the maintenance overhead of another language, the attendant cross-operating-system problems/cross-PyTorch-versions problems, and the issue of maintaining parity between two implementations. Obviously you feel differently though; that's fine.*

**Comment by lxuechen** [comment]
> Okay, so, plan is to have BPath, BTree be wrappers for BInterval, using the old API. This will result in minor backward incompatibilities (no .to) but should be fine. The main advantage being that there is now only have one Brownian* object to maintain, rather than five.

I mostly agree with this proposal. I don't think it hurts to support both how we previously constructed things and what we currently do, and we'll just state the recommendation is to follow the new API. 

> The speed regression is really just because of the C++ -> Python switch, not because of the choice to implement BPath via BInterval. As such I anticipate that it's only temporary. My suggestion for C++ going forwards is to implement just BrownianInterval.

I'm not entirely sure that's true. When queries exceed the cache, the reconstruction will take time, right? This is not to mention the various object constructions which will put quite some load on the interpreter. Though, I agree I hadn't taken the chance to fully optimize Bpath, and I agree that any speculation on performance not based on profiling doesn't mean much. My original point was to provide this as an option so that there's a chance to squeeze some performance out if that's ideal. 

So here's a proposal, we could do the wrapper logic, and I'll additionally add a `BPathv2` (open to recommendations for other names) implemented in C++. Maintaining parity between two implementations shouldn't be a problem, if the base class is well-defined. I'm happy to invest in the time to do this.

**Comment by patrick-kidger** [comment]
> I mostly agree with this proposal. I don't think it hurts to support both how we previously constructed things and what we currently do, and we'll just state the recommendation is to follow the new API.

The old and new APIs are different. It's not possible to have `BrownianPath(new stuff)` and `BrownianPath(old stuff)` both work. If we want to support the new API wrt path/tree then I'd probably do it via classmethods on Interval.

> I'm not entirely sure that's true. When queries exceed the cache, the reconstruction will take time, right? This is not to mention the various object constructions which will put quite some load on the interpreter. Though, I agree I hadn't taken the chance to fully optimize Bpath, and I agree that any speculation on performance not based on profiling doesn't mean much. My original point was to provide this as an option so that there's a chance to squeeze some performance out if that's ideal.
> 
> So here's a proposal, we could do the wrapper logic, and I'll additionally add a `BPathv2` (open to recommendations for other names) implemented in C++. Maintaining parity between two implementations shouldn't be a problem, if the base class is well-defined. I'm happy to invest in the time to do this.

Setting `cache_size=None` gives an infinite cache and no reconstruction. This looks to give performance roughly equivalent to the old BPath, although I've only done rough checks of that fact. (Not too surprising; BPath, BInterval and BTree all ultimately use a tree based data structure, either explicitly or with blist.)
I essentially regard BInterval as a hybridisation/generalisation of both BPath and BTree, which is why it's possible to specialise back to each of them, and why I suggest BInterval as the thing to implement in C++.

**Comment by patrick-kidger** [comment]
@lxuechen Returning to this now we all have a little time. Path and Tree now maintain the same interface (and pass the same tests) as before.

File: tests/test_brownian_interval.py [hunk 1]
```diff
@@ -38,12 +37,16 @@
 REPS = 3
 MEDIUM_REPS = 25
 LARGE_REPS = 500
-ALPHA = 0.0001
+ALPHA = 0.00001
```
Comment (lxuechen) [inline tests/test_brownian_interval.py:40]: Why did this significance level become even smaller?

File: torchsde/_brownian/derived.py [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+    @property
+    def shape(self):
+        return self.base_brownian.shape
+
+    @property
+    def levy_area_approximation(self):
+        return self.base_brownian.levy_area_approximation
+
+
+class BrownianPath(brownian_interval.BrownianInterval):
```
Comment (lxuechen) [inline torchsde/_brownian/derived.py:50]: Is it possible to use some form of dependency injection here, as opposed to inheriting directly from BI? BP should inherit from BaseB, since the latter specifies the API. But inheriting from BI could be a rather strong dependency.

File: torchsde/_brownian/derived.py [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+        # The extra tb, return_U, return_A arguments are all needed for compatibility with the current solvers.
+        assert not return_U
+        assert not return_A
+        out = super(BrownianPath, self).__call__(t, tb)
+        if tb is None:
+            out = out + self._w0
+        return out
+
+
+class BrownianTree(brownian_interval.BrownianInterval):
```
Comment (lxuechen) [inline torchsde/_brownian/derived.py:87]: A similar depency argument goes here.

File: torchsde/_brownian/derived.py [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+    To use:
+    >>> bm = BrownianPath(t0=0.0, w0=torch.zeros(4, 1))
+    >>> bm(0., 0.5)
+    tensor([[ 0.0733],
+            [-0.5692],
+            [ 0.1872],
+            [-0.3889]])
+    """
+
+    def __init__(self, t0: Scalar, w0: Tensor, window_size: int = 8):
```
Comment (lxuechen) [inline torchsde/_brownian/derived.py:64]: Could we also try support the new API at the same time? Something like `__init__(self, t0, t1, shape..., w0=None, window_size=...)`? The legacy `w0` and `window_size` should mostly be there to not anger people who might have code depending on this codebase.

File: torchsde/_brownian/derived.py [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+    To use:
+    >>> bm = BrownianTree(t0=0.0, w0=torch.zeros(4, 1))
+    >>> bm(0., 0.5)
+    tensor([[ 0.0733],
+            [-0.5692],
+            [ 0.1872],
+            [-0.3889]], device='cuda:0')
+    """
+
+    def __init__(self, t0: Scalar,
```
Comment (lxuechen) [inline torchsde/_brownian/derived.py:103]: Similar argument about constructor goes here.

**Review by lxuechen — COMMENTED** [review]
Thanks for bearing with my slowness during the last two weeks. 

In general the PR looks good. I'll add in the CPP version of BPath during the next weekend or two. 

Two additional comments I have is that:
- Could we switch the name of argument `shape` in constructor to `size`? This matches better with most torch Tensor construction functions (e.g. randn, zeros)? 
- Could we use lower case variable names for `W` and `H` in the constructor?

File: torchsde/_brownian/derived.py (hunk 2 above)
Comment (patrick-kidger) [reply torchsde/_brownian/derived.py:50]: Dependency injection, probably not. If you mean composition-over-inheritance, then yes that's possible - I was mostly just avoiding lots of redirections `def blah(self): return self._interval.blah()`, but I can change that.

File: torchsde/_brownian/derived.py (hunk 4 above)
Comment (patrick-kidger) [reply torchsde/_brownian/derived.py:64]: The problem then is matching up positional arguments. It ends up having to be something like `__init__(self, t0, w0, window_size, ...)`, with some complicated set of checks to try and figure out if they're using old or new style API - in the latter case then re-interpreting `w0` as `t1`, and so on.

Moreover there's no real need to do this - if someone wants to use the new-style API then they can just use BInterval directly.

**Comment by patrick-kidger** [comment]
Absolutely no problem, we've all had a lot on our plate recently.

- If you're going for a C++ implementation then I'd suggest replicating BInterval. As per this PR, BInterval generalises both BPath and BTree, and I don't think there's any performance to be gained by special-casing either of them.

Actually, I think the best thing to implement would just be `_Interval`. That's the hot loop here. Wiring in the C++ also then becomes particularly easy - it's then just the private `BrownianInterval`<->`_Interval` interface that we need to worry about.

- shape vs size: agreed.

- lower case W, H: can do, although for example I've already been using lower case `h` to represent a time delta. Generally I'm happy to break Python casing conventions in favour of following mathematical conventions, but if you don't like it then I could change these to `increment` and `space_time_levy_area` (or similar) respectively.

File: torchsde/_brownian/derived.py (hunk 2 above)
Comment (lxuechen) [reply torchsde/_brownian/derived.py:50]: Yes, composition-over-inheritance is what I mean. The motivation is the old-school argument that in the case of inheritance, changing BI's API would affect BPath and BTree.

**Comment by lxuechen** [comment]
> lower case W, H

I agree with the math over naming convention now.

**Comment by patrick-kidger** [comment]
All done I think.
- Put the significance level back.
- Other changes as discussed.

File: README.md [hunk 6]
```diff
@@ -7,19 +7,25 @@ Similar to [torchdiffeq](https://github.com/rtqichen/torchdiffeq), algorithms in
...
-
 ## Installation
+#### Command
+
 ```shell script
 pip install git+https://github.com/google-research/torchsde.git
 ```
 
+#### Requirements
+Python 3 and PyTorch >=1.6.0
```
Comment (lxuechen) [inline README.md:18]: I guess if `python_requires='~=3.6'`, we might as well state Python >= 3.6 here.

File: tests/test_adjoint.py [hunk 7]
```diff
@@ -35,16 +35,14 @@
 dtype = torch.get_default_dtype()
 
 ito_methods = {'milstein': 'ito', 'srk': 'ito'}
-stratonovich_methods = {'midpoint': 'stratonovich', 'log_ode': 'stratonovich'}
```
Comment (lxuechen) [inline tests/test_adjoint.py:38]: Is the test flaky with log_ode?

File: tests/test_adjoint.py [hunk 8]
```diff
@@ -35,16 +35,14 @@
...
 
 
 @pytest.mark.parametrize("sde_cls", [Ex1, Ex2, Ex3, Ex4])
 @pytest.mark.parametrize("method, sde_type", itertools.chain(ito_methods.items(), stratonovich_methods.items()))
 @pytest.mark.parametrize('adaptive', (False,))
 def test_adjoint(sde_cls, method, sde_type, adaptive):
     # Skipping below, since method not supported for corresponding noise types.
-    if method == METHODS.log_ode_midpoint and sde_cls.noise_type == NOISE_TYPES.diagonal:
-        return
     if method == METHODS.milstein and sde_cls.noise_type == NOISE_TYPES.general:
```
Comment (lxuechen) [inline tests/test_adjoint.py:46]: Optional: Not super important, but I guess we could merge these two if checks.

**Comment by lxuechen** [comment]
Thanks again for the shape->size change! Looking at the diff, the task seems like a slog.

**Comment by patrick-kidger** [comment]
- Tweaks as suggested.
- I removed the log-ODE test because it's quite slow: ~500 seconds, which is about 1/3 of the total runtime of the tests. On that note, I don't think we actually have tests that check that all the different solvers.
- shape->size was easy, my IDE did it for me!
- FYI I'm putting together a PR atm to update the README.

**Comment by lxuechen** [comment]
> shape->size was easy, my IDE did it for me!
